### PR TITLE
Preparing to use futures in storage.

### PR DESCRIPTION
**Pull request by dhermes**
Wraps setting/getting of object `_properties` in custom methods. This will allow centralized detection of a future in a response and will also allow replacing with the value on access if it is ready.

Towards #775 

----

@tseaver LMK what you think of this. The next step (I have yet to write it) is to make `Batch._do_request` return a future object. These objects will be owned and kept in order by the `Batch` and eventually populated. (This may be an issue because `JSONConnection.api_request` will try to parse the response if we use `expect_json=True`. We simultaneously want to parse the non-batched JSON responses and just ignore the futures)

### CODE REVIEW

File: gcloud/storage/blob.py [hunk 1]
```diff
@@ -358,7 +359,7 @@ def upload_from_file(self, file_obj, rewind=False, size=None,
         if not isinstance(response_content,
                           six.string_types):  # pragma: NO COVER  Python3
             response_content = response_content.decode('utf-8')
-        self._properties = json.loads(response_content)
+        self._set_properties(json.loads(response_content))
```
Comment (tseaver) [inline gcloud/storage/blob.py]: This looks like it fixes a bug where `self._changes` was not being cleared.

**Comment by tseaver** [comment]
The whole `self._properties` -> `self._get_properties()` dance seems to me to call for having the "future" bits hold a `_properties` which is a special object, whose `__getitem__`/`__setitem__`/`get` would raise the appropriate error.  Converting the future would then be done by replacing the `_properties` attribute with a real dict.

**Comment by dhermes** [comment]
1. I figured you'd have good ideas on this, so I pushed it out without thinking too hard about it.
1. I don't particularly like `_(get|set)_properties`
1. I think your suggestion is a bit too high-powered for what is happening. I prefer simplicity if possible.
1. Writing this and #812 I very much feel like this stuff is too intertwined (i.e. modules / classes know / assume too much about each other)

File: gcloud/storage/blob.py (hunk 1 above)
Comment (dhermes) [inline gcloud/storage/blob.py]: Indeed. My bad / good.

**Comment by tseaver** [comment]
I imagined the following:

```python
class UnrealizedFuture(ValueError):
    pass

class _UnrealizedFutureDict(object):

    def get(self, key, default=None):
        raise UnrealizedFuture()

    def __getitem__(self, key):
        raise UnrealizedFuture()

    def __setitem__(self, key, value):
        raise UnrealizedFuture()

unrealized = _UnrealizedFutureDict()
```

Where every "future" `Blob` / `Bucket` would have the `unrealized` singleton as its `_properties`.  That doesn't seem "high-powered" to me, compared to the ugliness of the `_get_properties()` bit scattered everywhere.

**Comment by dhermes** [comment]
That's much simpler than I imagined! Let me give it a spin.

**Comment by dhermes** [comment]
@tseaver I got rid of `_get_properties` in light of your suggestion but kept around `_set_properties` since it also resets `_changes` for us. Kept the old version of this PR around in https://github.com/dhermes/gcloud-python/tree/prepare-for-storage-futures-OLD for now, but will delete soon-ish.

I will give this a spin in #812 with your suggestion. It totally assuages my concerns about modules knowing too much about each other.

**Comment by tseaver** [comment]
LGTM

**Pull request merged**